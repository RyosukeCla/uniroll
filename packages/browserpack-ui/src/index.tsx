export { EnvContext, useEnv } from "./components/contexts";
export { App } from "./components/App";
