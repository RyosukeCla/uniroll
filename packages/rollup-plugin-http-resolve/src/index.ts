import { Plugin } from "rollup";
import path from "path";

export type ImportMaps = {
  imports: { [k: string]: string };
};

function isHttpProtocol(id: string) {
  return id.startsWith("http://") || id.startsWith("https://");
}

const DEBUG = false;
const log = (...args: any) => DEBUG && console.log(...args);

type HttpResolveOptions = {
  cache?: any;
  fetcher?: (url: string) => Promise<string>;
  onRequest?: (url: string) => void;
  onUseCache?: (url: string) => void;
  transform?: (code: string) => string;
  fallback?: (
    id: string,
    importer: string,
    onWarn: (warn: any) => void
  ) => Promise<string | void> | void | string;
};
const defaultCache = new Map();
export const httpResolve = function httpResolve_({
  cache = defaultCache,
  onRequest,
  transform,
  onUseCache,
  fetcher,
  fallback,
}: HttpResolveOptions) {
  return {
    name: "http-resolve",
    async resolveId(id: string, importer: string) {
      log("[http-resolve:resolveId:enter]", id, "from", importer);
      // on network resolve
      if (importer && isHttpProtocol(importer)) {
        if (id.startsWith("https://")) {
          log("[http-reslove:end] return with https", id);
          return id;
        }
        const { pathname, protocol, host } = new URL(importer);
        if (id.startsWith("/")) {
          // /_/...
          log(
            "[http-reslove:end] return with host root",
            `${protocol}//${host}${id}`
          );
          return `${protocol}//${host}${id}`;
        } else if (id.startsWith(".")) {
          // ./xxx/yyy
          // relative path
          const resolvedPathname = path.join(path.dirname(pathname), id);
          const newId = `${protocol}//${host}${resolvedPathname}`;
          log("[http-resolve:end] return with relativePath", newId);
          return newId;
        }
      }
      if (fallback) {
        const fallbacked = await fallback(id, importer, this.warn);
        log("[http-resolve:end] use fallback to", id, "=>", fallbacked);
        if (fallbacked) {
          return fallbacked;
        }
      }
    },
    async load(id: string) {
      log("[http-resolve:load]", id);
      if (id === null) {
        this.warn("irregular missing id");
        return;
      }
      if (isHttpProtocol(id)) {
        const cached = await cache.get(id);
        if (cached) {
          onUseCache?.(id);
          return cached;
        }
        onRequest?.(id);
        if (fetcher) {
          let code = await fetcher(id);
          code = transform?.(code) ?? code;
          await cache.set(id, code);
          return code;
        } else {
          const res = await fetch(id);
          if (!res.ok) {
            throw res.statusText;
          }
          let code = await res.text();
          code = transform?.(code) ?? code;
          await cache.set(id, code);
          return code;
        }
      }
    },
  } as Plugin;
};

export function createImportMapsFallback({
  importmaps,
}: {
  importmaps?: ImportMaps;
}) {
  return (
    id: string,
    importer: string | void = undefined,
    warn: (warning: any) => any
  ): Promise<void | string> | void | string => {
    if (importer == null) {
      return;
    }
    if (id.startsWith("http")) {
      return;
    }
    if (id.startsWith(".")) {
      return;
    }

    const mapped = importmaps?.imports[id];
    if (mapped) {
      return mapped;
    }
    warn(`missed fallback to https://cdn.skypack.dev/${id}`);
    return `https://cdn.skypack.dev/${id}`;
  };
}

// export default httpResolve;
