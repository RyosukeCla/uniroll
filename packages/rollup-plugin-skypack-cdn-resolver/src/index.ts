import { Plugin } from "rollup";
import path from "path";
const SKYPACK_CDN_HOST = "https://cdn.skypack.dev";

export function skypackCDNResolver({
  cache = new Map(),
  onRequest,
  onUseCache,
  ignorePolyfill = false,
}: {
  cache?: any;
  ignorePolyfill?: boolean;
  onRequest?: (url: string) => void;
  onUseCache?: (url: string) => void;
}) {
  return {
    async resolveId(id: string, importer: string) {
      // console.log(id, importer);
      if (importer && importer.startsWith(SKYPACK_CDN_HOST)) {
        // load Skypack in Skypack
        if (id.startsWith(SKYPACK_CDN_HOST)) {
          return id;
        }

        const { pathname, protocol, host } = new URL(importer);
        if (id.startsWith("/")) {
          return `${protocol}//${host}${id}`;
        } else {
          const resolvedPathname = path.join(path.dirname(pathname), id);
          const newId = `${protocol}//${host}${resolvedPathname}`;
          return newId;
        }
      }
    },
    async load(id: string) {
      if (id.includes("@pika/polyfill")) {
        return `// ignored: ${id}`;
      }
      if (id.startsWith(SKYPACK_CDN_HOST)) {
        const cached = await cache.get(id);
        if (cached) {
          onUseCache && onUseCache(id);
          return cached;
        }
        onRequest && onRequest(id);
        const res = await fetch(id);
        if (!res.ok) {
          throw res.statusText;
        }
        const code = await res.text();
        await cache.set(id, code);
        return code;
      }
    },
  } as Plugin;
}
