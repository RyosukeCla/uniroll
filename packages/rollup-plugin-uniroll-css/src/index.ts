import { Options } from "./types";
import { Plugin } from "rollup";
import { transformCss } from "./transformCss";
import { transformWithAutoprefixer } from "./transformWithAutoprefixer";
import { wrapWithStyleInjector } from "./wrapWithInjector";

export const css = (options: Options = {}) => {
  return {
    async transform(code: string, id: string): Promise<string | void> {
      if (id.endsWith(".css")) {
        const css = await transformCss(code, options);
        const autoprefixedCss = await transformWithAutoprefixer(css);
        return wrapWithStyleInjector(autoprefixedCss);
      }
    }
  } as Plugin;
};
