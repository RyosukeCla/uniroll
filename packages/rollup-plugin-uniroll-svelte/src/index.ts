import { PreprocessorGroup } from "svelte/types/compiler/preprocess";
import type { Plugin } from "rollup";
import { compile as svelteCompile, preprocess } from "svelte/compiler";
import {
  cdnRewriteTransformerFactory,
  createSveltePreprocessor,
} from "./tsPreprocess";
import ts from "typescript";
import { CompileOptions } from "svelte/types/compiler/interfaces";
import { ResolveIdFallback } from "./types";

type SveltePluginOptions = {
  target: ts.ScriptTarget;
  resolveIdFallback?: ResolveIdFallback;
  svelteOptions?: CompileOptions;
  extraPreprocess?: PreprocessorGroup[];
};

const defaultResolveIdFallback: ResolveIdFallback = (
  id: string,
  importer?: string
) => {
  return;
};

export const svelte = (opts: SveltePluginOptions) => {
  /**
   * if css is set to false, then output css file instead of js injection.
   */
  const willOutptuCss = !opts.svelteOptions?.css;
  const cssFiles = new Map<string, any>();

  return {
    name: "uniroll-svelte",
    buildStart() {
      cssFiles.clear();
    },
    load(id: string) {
      if (!cssFiles.has(id)) return;
      return cssFiles.get(id);
    },
    async transform(code: string, id: string) {
      if (id.endsWith(".svelte")) {
        const tsPreprocess = createSveltePreprocessor({
          target: opts.target,
          importer: id,
          resolveIdFallback: opts.resolveIdFallback ?? defaultResolveIdFallback,
        });

        const { code: preprocessed } = await preprocess(
          code,
          [tsPreprocess, ...(opts.extraPreprocess ?? ([] as any))],
          {
            filename: id,
          }
        );
        const result = svelteCompile(preprocessed, opts.svelteOptions);
        if (result.warnings.length > 0) {
          this.warn(result.warnings.map((t) => t.message).join("\n"));
        }

        if (willOutptuCss && result.css.code) {
          const cssFileName = `/svelte-generated-css/${Math.random().toString(16)}.css`;
          result.js.code += `\nimport "${cssFileName}";\n`;
          cssFiles.set(cssFileName, result.css);
        }

        const ret = ts.transpileModule(result.js.code, {
          fileName: "$temp.tsx",
          compilerOptions: {
            target: opts.target,
            module: ts.ModuleKind.ESNext,
          },
          transformers: {
            before: [cdnRewriteTransformerFactory(opts.resolveIdFallback!, id)],
          },
        });

        return {
          code: ret.outputText,
          map: ret.sourceMapText,
        };
      }
    },
  } as Plugin;
};
