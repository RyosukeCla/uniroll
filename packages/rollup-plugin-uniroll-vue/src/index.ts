import qs from "querystring";
import type { Plugin, TransformResult } from "rollup";
// import type { SFCAsyncStyleCompileOptions } from "@vue/compiler-sfc";

const cssInjectionMarker = "__UNIROLL_VUE_CSS__";
// const cssPreprocessLangRE = /\.(less|sass|scss|styl|stylus|postcss)$/;
// const cssModuleRE = /\.module\.(less|sass|scss|styl|stylus|postcss|css)$/;
// const isCSSRequest = (file: string) => file.endsWith(".css") || cssPreprocessLangRE.test(file);

export function unirollVue() {
  return {
    name: "uniroll-plugin-vue",
    async transform(code: string, id: string): Promise<TransformResult> {
      const isVueStyle = /\?vue&type=style/.test(id);
      if (isVueStyle) {
        // TODO: preprocessing other alt css lang
        // const preprocessLang = (id.match(cssPreprocessLangRE) || [])[1] as SFCAsyncStyleCompileOptions["preprocessLang"];
        // const query = parseVuePartRequest(id);
        const genCode =
          `let ${cssInjectionMarker} = document.createElement('style');` +
          `${cssInjectionMarker}.innerHTML = ${JSON.stringify(code)};` +
          `document.head.appendChild(${cssInjectionMarker});`;
        return {
          code: genCode,
          map: null, // TODO:
        };
      } else {
        return code;
      }
    },
  } as Plugin;
}

/**
 * borow from rollup-plugin-vue
 * https://github.com/vuejs/rollup-plugin-vue/blob/3222451614e8c5e4c1c3188bd6388285a25851ba/src/index.ts#L320-L372
 */

type Query =
  | {
      filename: string;
      vue: false;
    }
  | {
      filename: string;
      vue: true;
      type: "script";
      src?: true;
    }
  | {
      filename: string;
      vue: true;
      type: "template";
      id?: string;
      src?: true;
    }
  | {
      filename: string;
      vue: true;
      type: "style";
      index: number;
      id?: string;
      scoped?: boolean;
      module?: string | boolean;
      src?: true;
    }
  | {
      filename: string;
      vue: true;
      type: "custom";
      index: number;
      src?: true;
    };

function parseVuePartRequest(id: string): Query {
  const [filename, query] = id.split("?", 2);
  if (!query) {
    return { vue: false, filename };
  }

  const raw = qs.parse(query);
  if ("vue" in raw) {
    return {
      ...raw,
      filename,
      vue: true,
      index: Number(raw.index),
      src: "src" in raw,
      scoped: "scoped" in raw,
    } as any;
  }

  return { vue: false, filename };
}
