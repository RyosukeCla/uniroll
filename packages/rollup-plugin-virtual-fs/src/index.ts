import type { Plugin } from "rollup";

import path from "path";

const SEARCH_EXTENSIONS = [
  "/index.tsx",
  "/index.ts",
  "/index.js",
  ".tsx",
  ".ts",
  ".json",
  ".js",
];

function searchFile(
  vfs: Map<string, string>,
  filepath: string,
  extensions: string[]
) {
  for (const ext of ["", ...extensions]) {
    // console.log("searching...", filepath + ext);
    if (vfs.has(filepath + ext)) {
      return filepath + ext;
    }
  }
}

const DEBUG = false;
const log = (...args: any) => {
  if (DEBUG) console.log(...args);
};

const isFileSchema = (id: string) =>
  id.startsWith("file://") || id.startsWith("/");

const isRelativePath = (id: string) => stripSchema(id).startsWith(".");
const stripSchema = (id: string) => id.replace(/^file\:(\/\/)?/, "");

export const virtualFs = ({
  files,
  extensions = SEARCH_EXTENSIONS,
  memoryOnly = true,
}: {
  files: { [k: string]: string };
  extensions?: string[];
  memoryOnly?: boolean;
}) => {
  const vfs = new Map(Object.entries(files));
  return {
    name: "virtual-fs",
    async resolveId(id: string, importer: string | undefined) {
      // const exts = extensions ?;
      log("[rollup-plugin-virtual-fs]", id, importer);
      const normalized = stripSchema(id);
      // entry point
      if (isFileSchema(id) && importer == null) {
        return searchFile(vfs, normalized, extensions);
      }
      // relative filepath
      if (importer && isFileSchema(importer) && isRelativePath(id)) {
        const rawImporter = importer.replace(/^file\:/, "");
        const fullpath = rawImporter
          ? path.resolve(path.dirname(rawImporter), normalized)
          : id;
        const reslovedWithExt = searchFile(vfs, fullpath, extensions);
        if (reslovedWithExt) return reslovedWithExt;
        this.warn(`[rollup-plugin-virtual-fs] can not resolve id: ${fullpath}`);
      }
    },
    load(id: string) {
      const real = stripSchema(id);
      const ret = vfs.get(real);
      if (ret) return ret;
      if (memoryOnly)
        throw new Error(`[virtualFs] ${id} is not found on files`);
    },
  } as Plugin;
};
