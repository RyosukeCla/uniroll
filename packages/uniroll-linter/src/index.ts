import type { Linter as EslintLinter } from "eslint";
import Linter from "./Linter";

const linter = new Linter();

export function lint(
  code: string,
  filename?: string,
  config?: EslintLinter.Config
): EslintLinter.LintMessage[] {
  const messages = linter.verify(
    code,
    {
      parserOptions: {
        ecmaVersion: 2015,
        sourceType: "module"
      },
      env: {
        browser: true
      },
      globals: { module: true },
      rules: {
        "no-undef": "error"
      },
      ...config
    },
    { filename }
  );
  return messages;
}
