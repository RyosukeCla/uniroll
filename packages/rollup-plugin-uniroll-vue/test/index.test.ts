import { SourceDescription } from "rollup";
import { unirollVue } from "../src/";

test("rollup-plugin-vue parsed style query", async () => {
  const transform = unirollVue().transform as (
    code: string,
    id: string
  ) => Promise<SourceDescription>;
  const id = "/App.vue?vue&type=style&index=0&id=72300062&scoped=true&lang.css";
  const code = `p[data-v-72300062] { color: red; }`;
  const { code: genCode } = await transform(code, id);

  expect(genCode).toMatchSnapshot();
});

test("rollup-plugin-vue not parsed query", async () => {
  const transform = unirollVue().transform as (
    code: string,
    id: string
  ) => Promise<string>;
  const id = "/index.js";
  const code = `export default 1;`;
  const genCode = await transform(code, id);

  expect(genCode).toEqual(`export default 1;`);
});
