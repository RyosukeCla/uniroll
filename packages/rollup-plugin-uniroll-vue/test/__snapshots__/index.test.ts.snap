// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rollup-plugin-vue parsed style query 1`] = `"let __UNIROLL_VUE_CSS__ = document.createElement('style');__UNIROLL_VUE_CSS__.innerHTML = \\"p[data-v-72300062] { color: red; }\\";document.head.appendChild(__UNIROLL_VUE_CSS__);"`;
